Sprint 201808

sabicalija opened the pull request:
Implemented Sprint 201808:
- setup vue project and main application layout (nav, sidebar, content)
- integrate vuex, vue-router and vuetify
- implement ACS-Help content display (GitHub API)
- separate Mobile/PC navigation menu

sabicalija committed:
Implement vuex store, move component states to (vuex) store

sabicalija committed:
Minor corrections: duplicate keys in v-for, data funtions doesn't return object

sabicalija committed:
Migrate fontawesome icon loading from main.js into separate file (src/libraries)

sabicalija committed:
Migrate bootstrap from main.js into separate file (src/libraries)

sabicalija committed:
Remove underscore from previously wrong names subfolder (asterics-web-vue_)

sabicalija committed:
Add help entry (nav), add vuetify: add dropdown, implement tag (vX[.Y][.Z]) listing

sabicalija committed:
Minor fixes: eslint related

sabicalija committed:
Implement content loading (meta info) of git repository based on selected tag

sabicalija committed:
Update theme colors (vuetify: update, add error warning info success)

sabicalija committed:
asterics-web help: update version-button, add content-section

sabicalija committed:
asterics-web help: implement content-section, add and implement help-content (display selected help content, replace rel. img. path with abs.)

sabicalija committed:
Integrate vuetify, exchange Bootstrap + FontAwesome with Vuetify, add sidebar, remove unneccessary components, add 2nd route: sidebar

sabicalija committed:
Implement comoponent documentation, doc-sidebar, doc-content and recursive doc-element

sabicalija committed:
Implement component doc-tag-select, rate-limit

sabicalija committed:
Implement git tree loading, mutate flat array to tree functionality, update displayed element using new tree structure

sabicalija committed:
Remove file not needed anymore: help, help-content, content-section, version-button

sabicalija committed:
Remove rest of files and packages not used: Boostrap, FontAwesome; npm: @fortawesome/fontawesome-svg-code, @fortawesome/free-brands-svg-icons, @fortawesome/free-regular-svg-icons, @fortawesome/free-solid-svg-icons, @fortawesome/vue-fortawesome, bootstrap, bootstrap-vue

sabicalija committed:
Generate new package-lock.json

sabicalija committed:
Correct isFolder() checking directory name

sabicalija committed:
Implement global sidebar status using vuex, add link to home in navigation, remove nested sidebar

sabicalija committed:
Seperate navigation into 2 components (pc: navigation-toolbar, mobile: navigation-expansion), add css property  to pictures displayed in documentation section, add fab to mobile version (sidebar toggler)

deinhofer committed:
changed img-style property to 'max-width' to display images in their natural size but with a max-width of page width.

sabicalija committed:
Collapse navigation (mobile) after selecting new destination in the toolbar

deinhofer commented on asterics-web-vue/src/components/general/NavigationExpansion.vue:
```diff
@@ -61,7 +61,12 @@ export default {
     expansionPanelState(v) {
-      this.expansion = v;
+      // this.expansion = v;
+      if (v) this.expansion = 0;
```
why is the expansion state not boolean or simply 0/1?
why do you have to set it to null?

sabicalija replied:
Because `expansion` is "modeled" by the `<v-expansion-panel/>`, above.
The expansion panel controls this variable and sets it either to `0` or `null`.

```vue
<v-expansion-panel v-model="expansion" dark>
  <v-expansion-panel-content class="primary" @input="expansionPanelState($event)">
```

As a matter of fact, I was surprised too. :neutral_face: 

Long story short: [`<v-expansion-panel/>`](https://vuetifyjs.com/en/components/expansion-panels) set's it to either `0` or `null`.

Setting the variable `expansion` to `0` and `1` does not correspond with the semantics in this case. In this case, `0` means that the `v-expansion-panel` is opened and `null` that it is collapsed.

---

Variables bond by the [`v-model`](https://vuejs.org/v2/api/#v-model) directive are used in vue to create a so-called [two-way data binding](https://vuejs.org/v2/guide/forms.html).


In our case, the variable `expansion` is updated by the `v-expansion-model` and by `expansionPanelState(v)`. When the navigation is changed by the user, it is updated by the respective component (i.e. `<v-expansion-panel/>`) when an "internal value" changes.

At the same time, when the variable `expansion` is changed, let's say by a function (e.g. `expansionPanelState(v)`), the "internal value" is updated by `v-model` too (data flow in opposite direction).

klues commented on asterics-web-vue/package.json:
```diff
@@ -15,19 +15,14 @@
     "vuex": "^3.0.1"
   },
   "devDependencies": {
+    "@fortawesome/fontawesome-free": "^5.2.0",
```
why fontawesome is only a dev-dependency now?

sabicalija replied:
I've forgot to remove them. Thanks!

sabicalija replied:
Sorry. I was wrong.

This dependency (fontawesome-free) still remains, but I'm a little bit confused about what is considered a development dependency and what a normal one. Do you mind explaining it?

deinhofer replied:
as I understand, the package is only needed during the build process, but not for the final result. But I can't imagine that a font would not be needed in the deployment, unless it is substituted by an svg or so.

sabicalija replied:
I had the same idea in mind. But do we need the components listed at `dependencies` or could they be moved to `devDependencies` too?

klues replied:
yes, it is like @deinhofer said. `babel` is a typical `dev-dependency`, because it is only needed for building / transpiling, but not at runtime. `vue` is a typical `dependency`, because it is needed at runtime (runs in the browser). the only thing that remains where I'm not sure if it is a true `dependency` is `register-service-worker`. Do you know what it is?

sabicalija replied:
The project created and some of the dependencies (incl. `register-service-worker`) were added by `vue-cli`. I do not know, what it is for.

klues commented on asterics-web-vue/src/components/documentation/DocContent.vue:
```diff
@@ -0,0 +1,66 @@
+    <v-container fluid>
+      <v-layout row wrap>
+        <v-flex xs12>
+          <span v-html="content"></span>
```
A`<div>` element probably would be more correct here than a `<span>`. Spans are normally small inline elements and here its the parent of a bunch of html that will be rendered inside, isnt it?

sabicalija replied:
:+1:

klues commented on asterics-web-vue/src/components/documentation/DocContent.vue:
```diff
@@ -0,0 +1,66 @@
+    };
+  },
+  computed: {
+    ...mapGetters(["helpInfo"])
```
whats this `...` syntax? I do not know it.

sabicalija replied:
It's called "spread syntax" (cf. [here](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Operators/Spread_syntax)). You can use it to kind of "expand" arrays.
But I did not research it to any extend. It's what I've found in the vue.js docs.

klues replied:
ah, thx. did not know it. just confused it with the java syntax for an arbitrary number of arguments, https://stackoverflow.com/questions/7607353/how-to-create-java-method-that-accepts-variable-number-of-arguments

klues replied:
i am a little bit confused about your `computed` method. normally you define a property that is computed, like e.g. `reversedMessage` in the basic example in the docs: https://vuejs.org/v2/guide/computed.html But what property you do "compute" here and where do you use it?

sabicalija replied:
In many languages it is used for variadic parametrization. For instance C++ uses if for variadic parameter (incl. C) and variadic templates ([https://en.wikipedia.org/wiki/Variadic_template](https://en.wikipedia.org/wiki/Variadic_template)) :confused:.

---

The `helpInfo` variable is stored in the vuex store.
In the vuex docs, using data from vuex in a compontent, requires the use of mapGetters (which is listed in the computed object, in the docs - cf. [here](https://vuex.vuejs.org/guide/getters.html)).

klues replied:
ah, ok. so by this line you have all members of `helpInfo` from the store available in the local `this` namespace, correct?

sabicalija replied:
Correct :+1:

klues commented on asterics-web-vue/src/components/documentation/DocContent.vue:
```diff
@@ -0,0 +1,66 @@
+<template>
```
is this the module for including the existing asterics html-help files?

sabicalija replied:
yes

klues commented on asterics-web-vue/src/components/documentation/DocContent.vue:
```diff
@@ -0,0 +1,66 @@
+    let filename = this.$route.params[2];
+
+    // eslint-disable-next-line
+    let s = `${this.helpInfo.raw}${tag}/${this.helpInfo.path}/${path}/${filename}`;
```
let basePath = ``${this.helpInfo.raw}${tag}/${this.helpInfo.path}/${path}/``; let s = basePath + filename;

sabicalija replied:
Way better. Thanks! :+1:

klues commented on asterics-web-vue/src/components/documentation/DocContent.vue:
```diff
@@ -0,0 +1,66 @@
+        // eslint-disable-next-line
+        let c = html.replace(
+          r,
+          `src="${this.helpInfo.raw}${tag}/${this.helpInfo.path}/${path}/img`
```
`` `src=${basePath}img` ``

klues commented on asterics-web-vue/src/components/documentation/DocContent.vue:
```diff
@@ -0,0 +1,66 @@
+        );
+
+        r = new RegExp(`<img`, "g");
+        html = c.replace(r, `<img style="max-width: 100%;"`);
```
maybe below in the `<style>` element a css rule is nicer: `img { max-width: 100% }`?

sabicalija replied:
It's not possible in this case, unfortunately.
I've tried to add a class to those images instead.

```javascript
html = c.replace(r, `<img class="help-images"`);
```

and to add those property using this css class:

```css
.help-images{
  max-width: 100%;
}
```

But apparently it is not working. 
I would guess, that vue is not able to detect the dependencies in this case.
But first tests I've done so far, indicate that this is not the case.

klues replied:
ok, what a pity that it isnt working.

klues commented on asterics-web-vue/src/components/documentation/DocContent.vue:
```diff
@@ -0,0 +1,66 @@
+        r = new RegExp(`<img`, "g");
+        html = c.replace(r, `<img style="max-width: 100%;"`);
+
+        console.log(html);
```
in the end this should be removed

sabicalija replied:
:+1:

klues commented on asterics-web-vue/src/components/documentation/DocContent.vue:
```diff
@@ -0,0 +1,66 @@
+        // Replace pictures
+        let r = new RegExp(`src="img`, "g");
+        // eslint-disable-next-line
+        let c = html.replace(
```
why introducing a new variable `c`? why not `html = html.replace(...)`?

sabicalija replied:
Initially, I wasn't sure if this would be a good idea. But I'll change it. Thanks!

klues commented on asterics-web-vue/src/components/documentation/DocContent.vue:
```diff
@@ -0,0 +1,66 @@
+        console.log(html);
+
+        //FIXME: (implement) Change href
+        r = /href="/;
```
``html = html.replace(/href="/g, `href="${basePath}`)``

sabicalija replied:
Nice try. :wink:

This results in a link to rawgithub.com and users get only raw HTML files displayed.
Additionally, there are also links, which are actually valid, and no `basePath` needs to be added.

I'll address this issue in future sprints.

klues commented on asterics-web-vue/src/components/documentation/DocContent.vue:
```diff
@@ -0,0 +1,66 @@
+        let p = new DOMParser();
+        let d = p.parseFromString(html, "text/html");
+
+        this.content = d.body.innerHTML;
```
isnt `this.content = html` enough? I think the `v-html` in the template should automatically do the html rendering, also see: https://forum.vuejs.org/t/how-do-i-make-an-html-tag-inside-a-data-string-render-as-an-html-tag/13074

sabicalija replied:
I only want the body of the HTML page, received. That's why I have to parse the `html` variable, to extract the body of the page in `html` (`d.body.innerHTML`).

`v-html` renders HTML content automatically, indeed.

klues replied:
ah, ok, makes sense. if you only want to have the body, you probably have to do it this way.

klues commented on asterics-web-vue/src/components/documentation/DocElement.vue:
```diff
@@ -0,0 +1,77 @@
+<template>
```
what does this element do? i suppose I should see it in documentation? however locally none of these work on my computer: http://prntscr.com/kr7h6m

klues replied:
ok, I think I found it under `Support -> Help`. If the category on the page is called `Support` maybe also the folder containing the files for this category should be called `Support`.

sabicalija replied:
Well, I'm pretty sure, that the first image was made on an earlier branch, probably `sprint201808-doc`.

The element DocElement is used in the DocSidebar, and it uses some features that are rather discreet and won't be seen at first sight (see 2).

(1) I'll recommend you to install Vue Debugger Tools for Chrome. This way you'll find components like in this case, typically quickly.

![image](https://user-images.githubusercontent.com/13609768/45306027-e83f3100-b51b-11e8-91e2-4249a4475cb4.png)


(2) The component `DocSidebar` is actually a recursive component (cf. [here](https://vuejs.org/v2/guide/components-edge-cases.html#Recursive-Components)).
I've renamed the component unintentionally to `doc` in the source code. It should be more comprehensible now:

BEFORE:
```html
      .....
      <div v-for="i in element.children" :key="i.sha">
        <doc :element="i" :tag="tag"/>
      </div>
    </v-list-group>
  </div>
</template>
```
```javascript
import { mapGetters } from "vuex";

export default {
  name: "doc", // recursively used component <doc/>
  computed: {
```

NOW:
```html
      .....
      <div v-for="i in element.children" :key="i.sha">
        <doc-element :element="i" :tag="tag"/>
      </div>
    </v-list-group>
  </div>
</template>
```
```javascript
import { mapGetters } from "vuex";

export default {
  name: "doc-element",
  computed: {
```

klues replied:
ah, ok, thx. Then I indeed looked at the wrong version, I think it was the branch `sprint201808-doc`.
Yes, `doc-element` is better! Maybe `doc-sidebar-element` would be even better because then it would be very clear where this element is used. However it is also fine as it is.

sabicalija replied:
`doc-sidebar-element` is indeed better. I'll fix it in upcoming sprints. :+1:

klues commented on asterics-web-vue/src/components/documentation/DocSidebar.vue:
```diff
@@ -0,0 +1,96 @@
+      this.tag = v;
+
+      /* push new route if necessary */
+      // eslint-disable-next-line
```
what the thing eslint does not like?

sabicalija replied:
The length of the line. I did not figure out yet, how to change the configuration of "@vue/prettier".

klues commented on asterics-web-vue/src/components/documentation/DocElement.vue:
```diff
@@ -0,0 +1,77 @@
+      let r = /(.*)\/(.*)/.exec(path);
+
+      if (r == null) foldername = path;
+      else foldername = r[2];
```
why r[2]?

sabicalija replied:
The regex `r` above uses so-called *capturing groups*.

Every occurrence of `( ... )` is a capturing group.
If the match succeeds, I just want to store the 2nd capturing group in `foldername`.

klues commented on asterics-web-vue/src/components/documentation/DocElement.vue:
```diff
@@ -0,0 +1,77 @@
+      let str = i.replace(new RegExp("_", "g"), " ");
+
+      /* captialize */
+      return str.charAt(0).toUpperCase() + str.slice(1);
```
some dark magic is going on in lines 44-57 ;) I am sure it could be done somehow more readable.

sabicalija replied:
:bowtie: I'm not sure, how to change regex, to be more readable... 

I'll split the lines like:

```javscript
let r = /(.*)\/(.*).exec(r);
```

```
let r = new RegExp("(.*)\/(.*)");
let res = r.exec(s);
```

But it does not look more readable, actually.
Any suggestions?

I'll could add some comments to describe the regex, but I would not want to remove regexp's generally.

klues replied:
ok, I just did not (and still do not) get it what the whole method does. maybe you just should explain it to me next week and then I maybe have an idea to make it more readable.

deinhofer replied:
Maybe link to the page where you created the regex?
https://regex101.com/r/lsLAls/2

This let's you easily discover allowed and disallowed strings. So only the last part (filename) of a URL will be extracted. There is just one case where this does not work. If there are no leading '/'. Do we need to support this?

https://regex101.com/r/tIZCnk/1

sabicalija replied:
I've cleaned up the mess I created.. :relieved:

Please take a look.
I hope its clearer and easier to understand.
The first implementation was just quick-and-dirty to see if everything works.

Thanks to @deinhofer's idea using generated regex101.com links, no question should be left.

klues replied:
looks good!

klues commented on asterics-web-vue/src/libraries/util/help.js:
```diff
@@ -0,0 +1,90 @@
+  return await getS(l);
+};
+
+const getS = async url => {
```
what is `getS()`?

sabicalija replied:
not a good name, obviously. :grin:

It was short for "getStructure" or the like, but it will be or should be removed, as it is not necessary anymore. It was used to pull information of Folder from GitHub API, recursively.

Using git trees, however, recursion is not necessary anymore.

klues replied:
ok, then lets remove it. every bit of removed code is good - the shorter the better ;)

klues commented on asterics-web-vue/src/libraries/util/help.js:
```diff
@@ -0,0 +1,90 @@
+  for (let i = 0; i < array.length; i++) [await cb(array[i], i, array)];
+}
+
+const getContent = async l => {
```
is there a special reason for using `async` and `await`? browser compatibility is not the best and does it has any advantage?

sabicalija replied:
I'm not sure 100% of all pros and cons. I've used it currently because the code gets more readable, and so-called "callback hell" is avoided (cf. [here](https://blog.hellojs.org/asynchronous-javascript-from-callback-hell-to-async-and-await-9b9ceb63c8e8)).

---

The current project setup (vue-cli, webpack) should actually transform cases where *browser compatibility* issues could arise.

Configuration from `package.json` is used to ensure targeted browser support.

```javascript
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "not ie <= 8"
  ]
```

---

I did not have enough time so far, but with following two commands, one can analyse (or read at least) the generated code, either for older browsers, or newer ones --- and verify if the generated code includes any `async`/`await` or if they are replaced by webpack or babel-loader, respectively.

- `npm run build:modern`
- `npm run build:devel`

A combination of these commands (cf. `package.json`) could be used to generate _readable_ code, generated for modern and older browsers.

klues replied:
ok, until now I never used `async` and `await`, so I'm not sure either. You are right, the code maybe gets more readable, but at the same time maybe also slower, because everything is done sequential, while some things maybe could be done in parallel using callbacks and Promises.
I do not know if it is possible to make `async` and `await` compatible with older browsers at all because they are completeley new things that cannot be fixed with polyfills easily.

Yeah, but would be interesting what happens to the `async` calls in the generated code.

deinhofer commented:
Please fix change requests and then merge.

sabicalija committed:
Cleanup messy code, integrate feedback from PR:

 - use expressive names
 - tidy vuex store: move data to webapp, github
 - correct name of recursive component: doc-element
 - cleanup regex handling code, add regex101.com examples
 - move and add github functions from help.js to github.js

sabicalija committed:
Update @fortawesome/fontawesome-free package

sabicalija commented:
I've cleaned up the ugliest parts of the code:

 - use expressive names
 - tidy vuex store: move data to webapp, github
 - correct name of recursive component: doc-element
 - cleanup regex handling code, add regex101.com examples
 - move and add github functions from help.js to github.js

If you have time, take a look ([here](https://github.com/asterics/asterics-web/compare/dc99ae90a9d2513b5d9d788c8bca772557f077b8...master)) and give me feedback. :relaxed: 
@deinhofer @klues 
Thanks so much on the feedback so far! :+1:

sabicalija commented:
:tada: This PR is included in version 1.0.0 :tada:

The release is available on [GitHub release](https://github.com/asterics/asterics-web/releases/tag/v1.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

sabicalija commented:
:tada: This PR is included in version 1.0.0 :tada:

The release is available on [GitHub release](https://github.com/asterics/asterics-web/releases/tag/v1.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

sabicalija commented:
:tada: This PR is included in version 1.0.0 :tada:

The release is available on [GitHub release](https://github.com/asterics/asterics-web/releases/tag/v1.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
